File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    // ListNode* reverseList(ListNode* head) {
    //     if(head == NULL || head -> next == NULL) return head;

    //     ListNode* secondHead = reverseList(head->next);

    //     head->next->next = head;
    //     head->next = NULL;
    //     head = secondHead;

    //     return head;
    // }

    ListNode* reverseList(ListNode* head) {
        ListNode* prev = NULL;
        ListNode* curr = head;

        while(curr){
            ListNode* temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }

        head = prev;
        return head;
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {

        // 1. Find middle
        ListNode slow = head;
        ListNode fast = head;

        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }

        // 2. Split list
        ListNode secondHead = slow.next;
        slow.next = null;

        // 3. Reverse second half
        ListNode current = secondHead;
        ListNode prev = null;

        while(current != null){
            ListNode next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }

        secondHead = prev;

        // 4. Merge alternately
        ListNode first = head;

        while(secondHead != null){
            ListNode t1 = first.next;
            ListNode t2 = secondHead.next;

            first.next = secondHead;
            secondHead.next = t1;

            first = t1;
            secondHead = t2;
        }
    }
}